app/src/main/java/com/example/instasnap/Register.java

```diff
@@ -0,0 +1,164 @@
+
+    private EditText _editTextEmail;
+    private EditText _editTextPassword;
+    private Button _registerButton;
+    private FirebaseAuth _mAuth;
+    private ProgressBar _registerProgressBar;
+    private TextView _loginHereTextView;
+
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_register);
+
+        initialize();
+
+        setLoginNowButtonListener();
+
+        setRegisterButtonListener();
+
+    }
+
+    private void initialize() {
+        // Initializing registration objects
+        _mAuth = FirebaseAuth.getInstance();
+        _editTextEmail = findViewById(R.id.register_email_text_input);
+        _editTextPassword = findViewById(R.id.register_password_text_input);
+        _registerButton = findViewById(R.id.register_button);
+        _registerProgressBar = findViewById(R.id.register_progressBar);
+        _loginHereTextView = findViewById(R.id.login_now);
+    }
+
+    @Override
+    public void onStart() {
+        super.onStart();
+        // Check if user is signed in (non-null) and update UI accordingly.
+        FirebaseUser currentUser = _mAuth.getCurrentUser();
+        if(currentUser != null){
+            Intent homepageIntent = new Intent(getApplicationContext(), HomePage.class);
+            startActivity(homepageIntent);
+            finish();
+        }
+    }
+
+    @Override
+    public void onBackPressed() {
+        Intent loginIntent = new Intent(getApplicationContext(), Login.class);
+        startActivity(loginIntent);
+        finish();
+    }
+
+    private void setLoginNowButtonListener() {
+        // Anonymous Variable Example: One of the two listener type implementations requested
+        _loginHereTextView.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                Intent loginIntent = new Intent(getApplicationContext(), Login.class);
+                startActivity(loginIntent);
+                finish();
+            }
+        });
+    }
+
+
+    // Member Class Example: One of the two listener type implementations requested
+    private class AuthCompleteListener implements OnCompleteListener<AuthResult> {
+        private final Register _registerActivity;
+
+        public AuthCompleteListener(Register activity) {
+            this._registerActivity = activity;
+        }
+
+        @Override
+        public void onComplete(@NonNull Task<AuthResult> task) {
+            _registerActivity.onAuthenticationComplete(task);
+        }
+    }
+
+    public void onAuthenticationComplete(Task<AuthResult> task) {
+        _registerProgressBar.setVisibility(View.GONE);
+
+        if (task.isSuccessful()) {
+            Toast.makeText(Register.this, "Account created", Toast.LENGTH_SHORT).show();
+        } else {
+            // Registration failed, handle the error
+            try {
```

Comment:
This seems stupid, why throw an exception just to catch it right away.
Is there a way to use switch case here?

Reply:
If you read the exceptions, its to indicate what type missing information is required from the user, whether its a weak password, used email, invalid email or random error cause in the registration phase

Reply:
Firebase has made exceptions indicators to know exactly what was wrong in the registration by the user, hence the reason for using the throw and THEN catch it

Reply:
yes i understood the reason but isn't there another indicator without using the try catch?

**maybe something like:**

switch(task.getException()){
    catch FirebaseAuthWeakPasswordException:
              String errorMessage = "Weak password";
              Toast.makeText(getApplicationContext(), errorMessage, Toast.LENGTH_LONG).show();
     catch FirebaseAuthInvalidCredentialsException :
     .
     .
     .
 }